application/app/Models/Product.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Support\Facades\Storage;
+use Illuminate\Http\UploadedFile;
+use App\Models\ProductReview;
+use App\Models\ProductCategory;
+
+
+class Product extends Model
+{
+    /**
+     * @var array
+     */
+    protected $fillable = [
+        'image_path',
+        'name',
+        'price',
+        'description',
+        'product_category_id',
+    ];
+
+    /**
+     * @return \Illuminate\Database\Eloquent\Relations\HasMany
+     */
+    public function reviews()
+    {
+        return $this->hasMany(ProductReview::class);
+    }
+
+    /**
+     * @return \Illuminate\Database\Eloquent\Relations\BelongsTo
+     */
+    public function category()
+    {
+        return $this->belongsTo(ProductCategory::class, 'product_category_id', 'id');
+    }
+
+    /**
+     * @param $query
+     * @param string $id
+     * @return mixed
+     */
+    public function scopeEqualCategory(Builder $query, ?string $id)
+    {
+        if (is_null($id) || $id === "all") {
+            return;
+        }
+        return $query->where('product_category_id', $id);
+    }
+
+    /**
+     * @param $query
+     * @param string $name
+     * @return mixed
+     */
+    public function scopePartialName(Builder $query, ?string $name)
+    {
+        if (is_null($name)) {
+            return;
+        }
+        return $query->where('name', 'like', "%{$name}%");
+    }
+
+    /**
+     * @param $query
+     * @param string $price
+     * @return mixed
+     */
+    public function scopeWherePrice(Builder $query, ?string $price, ?string $price_compare)
+    {
+        if (is_null($price)) {
+            return;
+        }
+        if ($price_compare === "gteq") {
+            return $query->where('price', '>=', $price);
+        } else if ($price_compare === "lteq") {
+            return $query->where('price', '<=', $price);
+        }
+
+    }
+
+    /**
+     * @var array
+     */
+    private const SORT_COLUMN = [
+        'id',
+        'product_category',
+        'name',
+        'price',
+    ];
+
+    /**
+     * @var array
+     */
+    private const SORT_DIRECTION = [
+        'asc',
+        'desc',
+    ];
+
+    /**
+     * @param $query
+     * @param string $sort_column
+     * @param string $sort_direction
+     * @return mixed
+     */
+    public function scopeSortOrderBy(Builder $query, ?string $sort_column, ?string $sort_direction)
+    {
+        if (in_array($sort_column, self::SORT_COLUMN, true) &&
+            in_array($sort_direction, self::SORT_DIRECTION, true)) {
+            if ($sort_column === 'product_category') {
+                return $query->join('product_categories', 'product_category_id', 'product_categories.id')
+                    ->orderBy('product_categories.order_no', $sort_direction);
```

Comment:
この書き方テーブル明示してないから両方のテーブルに同じカラムあったら死ぬで
あとproduct_categoriesのカラムはほしくない(無駄な処理時間がかかる)のでselectで絞る
```suggestion
                return $query->join('product_categories', 'products.product_category_id', '=', 'product_categories.id')
                    ->select('products.*')
                    ->orderBy('product_categories.order_no', $sort_direction);
```
あと設計書の第２オーダー条件が抜けとる

Reply:
はい